modules/dnn/src/layers/convolution_layer.cpp

```diff
@@ -438,6 +438,18 @@ class ConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl
 #ifdef HAVE_OPENCL
         convolutionOp.release();
 #endif
+        if (inputs[0].dims == 3)
+        {
+            //Pool1D
+            #define REMOVE_EXTRA_ELEM(vector)\
+            if (vector.size() > 1)\
+                vector.pop_back();
+            REMOVE_EXTRA_ELEM(kernel_size);
+            REMOVE_EXTRA_ELEM(strides);
```

Comment:
Why we should drop part of information?
All these fields are initialized in ctor `BaseConvolutionLayerImpl()`.
This `finalize()` method can be called multiple times.
On each `finalize()` call such fields should have the same state on entry, otherwise we should have code to handle different states (which is hard to test, so we must avoid modification of "initial layer state").
It is unlikely that such network would keep alive during `setUp()` recall with updated parameters or input reshape.

BTW, `finalize()` is not the last stage of DNN "pipeline" before `forward*()`.

Reply:
> Why we should drop part of information?

Because it was unnecessarily added in https://github.com/opencv/opencv/blob/dd9bb3f2da961b23143705a3455a1968cbf63f5f/modules/dnn/src/layers/layers_common.cpp#L84
This code in ```layers_common.cpp``` is reached only when we load Caffe/Darknet networks or create OpenCV layers directly (like in ```test_layers.cpp```)

This code will work only once and will delete unnecessary elements from these vectors

Reply:
But does it works for Conv1D on CPU?

Reply:
> But does it works for Conv1D on CPU?

Yes, it works, but for CUDA I need these changes because of checks like this https://github.com/opencv/opencv/blob/0293e883e5cbcd99a3e31facfab6785b3604abf4/modules/dnn/src/cuda4dnn/primitives/convolution.hpp#L96

Reply:
So maybe we can move this closer to CUDA related part?

Reply:
Dilations are defined for the spatial dimensions only. So logically, the number of dilations must be equal to the number of spatial dimensions (which is equal to the convolution order). I think this assertion should stay as it is correct and this problem should be dealt elsewhere.

I think that resize is unnecessary (**and is probably a bug**). Somebody might have added it when OpenCV DNN supported 2D operations only (and it wasn't updated when 3D operations were introduced). It's presumably there for convenience so that user can specify a single value for all spatial dimensions (same padding for all dimensions for example). But clearly, this is wrong for other dimensions. It's wrong for 1D as you needed to patch it using `REMOVE_EXTRA_ELEM`. It's also wrong for 3D. What if you needed 3D convolution and provided a single value for padding? Resizing to two might tip off some assertions.

If fixing that isn't possible, I think `initCUDA` is a more appropriate place to do this. The whole `REMOVE_EXTRA_ELEM` is an attempt to make parameters stored in `BaseConvolutionalLayer` compatible with the constructor  of `cuda4dnn::ConvoltuionOp`. Hence, this code should be at the site where the `BaseConvolutionalLayer` parameters are transferred to `ConvolutionOp` (which is `initCUDA`).

Reply:
Note that preprocessor symbols do not have scope. The `REMOVE_EXTRA_ELEM` spills out of the scope where it was defined. I think you forgot an `#undef REMOVE_EXTRA_ELEM`.

Reply:
This is a bug. It was introduced in the 3D convolution PR (https://github.com/opencv/opencv/pull/14301). The intention was to broadcast parameters to all spatial dimensions if only one value was provided (the intent is mentioned in the review comments). The number of spatial dimensions was hardcoded to two for some reason.

**MCVE:**

```
#include <opencv2/dnn.hpp>

int main ()
{
    cv::dnn::Net net;

    {
        cv::dnn::LayerParams lp;

        int ksizes[] = {3};
        lp.set("kernel_size", cv::dnn::DictValue::arrayInt(ksizes, std::size(ksizes)));

        lp.set("num_output", 10);
        lp.set("bias_term", false);
        lp.type = "Convolution";
        lp.name = "testConv";

        int weightsShape[] = {10, 3, 3, 3, 3};
        cv::Mat weights(5, &weightsShape[0], CV_32F);
        randu(weights, -1.0, 1.0);
        lp.blobs.push_back(weights);

        net.addLayerToPrev(lp.name, lp.type, lp);
    }

    int inputShape[] = {1, 3, 5, 5, 5};
    cv::Mat input(5, &inputShape[0], CV_32F);
    randu(input, 1.0f, 2.0f);

    net.setInput(input);
    net.forward();
    return 0;
}
```

I was getting an floating point exception in `getMemoryShapes`(presumably due to garbage float in the OOB access`). So I added the following assertion:

```diff
 for (int i = 0; i < inpShape.size(); i++)
 {
+       CV_CheckLT(i, (int)kernel_size.size(), "OOB access in kernel size");
         outShape.push_back((inpShape[i] + pads_begin[i] + pads_end[i] - dilations[i] * (kernel_size[i] - 1) - 1) / strides[i] + 1);
 }
```

**Output:**

```
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.5.1-pre) /fakepath/opencv/opencv4/modules/dnn/src/layers/convolution_layer.cpp:371: error: (-2:Unspecified error) in function 'virtual bool cv::dnn::ConvolutionLayerImpl::getMemoryShapes(const std::vector<std::vector<int> >&, int, std::vector<std::vector<int> >&, std::vector<std::vector<int> >&) const'
> OOB access in kernel size (expected: 'i < (int)kernel_size.size()'), where
>     'i' is 2
> must be less than
>     '(int)kernel_size.size()' is 2

Aborted (core dumped)
```